### PR TITLE
Cleanup asserts & throws - replace with invariants - dir util/s* (part1)

### PR DESCRIPTION
**Pull request by sonodtt**
Cleanup of asserts - replaced with invariants.

Removed all unstructured `throw(...)` statements, replacing with `INVARIANT`, `PRECONDITION`, etc. where appropriate.

### CODE REVIEW

File: src/util/simplify_expr.cpp [hunk 1]
```diff
@@ -488,13 +488,13 @@ bool simplify_exprt::simplify_typecast(exprt &expr)
...
           expr=from_integer(1, expr_type);
-          assert(expr.is_not_nil());
+          CHECK_RETURN(expr.is_not_nil());
           return false;
         }
         else if(operand.is_false())
         {
-          expr=from_integer(0, expr_type);
-          assert(expr.is_not_nil());
+          expr = from_integer(0, expr_type);
```
Comment (tautschnig) [inline src/util/simplify_expr.cpp]: Unrelated change

File: src/util/simplify_expr.cpp [hunk 2]
```diff
@@ -1373,7 +1373,7 @@ bool simplify_exprt::simplify_update(exprt &expr)
       std::size_t number=to_struct_type(value_ptr_type).
         component_number(component_name);
 
-      assert(number<value_ptr->operands().size());
+      CHECK_RETURN(number < value_ptr->operands().size());
```
Comment (tautschnig) [inline src/util/simplify_expr.cpp]: This code should be rewritten to use `get_component`.

File: src/util/simplify_expr.cpp [hunk 3]
```diff
@@ -1409,7 +1409,9 @@ bool simplify_exprt::simplify_object(exprt &expr)
   {
     const typet &op_type=ns.follow(expr.op0().type());
 
-    assert(expr.operands().size()==1);
+    DATA_INVARIANT(
+      expr.operands().size() == 1,
+      "typecasts must have exactly one argument");
```
Comment (tautschnig) [inline src/util/simplify_expr.cpp]: Introduce `const typecast_exprt &tc = to_typecast_expr(expr);` above and change the code to use `tc` with its proper API. Let's not duplicate consistency checks on expressions.

File: src/util/simplify_expr_boolean.cpp [hunk 4]
```diff
@@ -245,7 +245,9 @@ bool simplify_exprt::simplify_not(exprt &expr)
   }
   else if(op.id()==ID_exists) // !(exists: a) <-> forall: not a
   {
-    assert(op.operands().size()==2);
+    DATA_INVARIANT(
+      op.operands().size() == 2,
+      "exists expression has two parameters");
```
Comment (tautschnig) [inline src/util/simplify_expr_boolean.cpp]: Please introduce `to_exists_expr` which should do this sort of check and use it here.

File: src/util/simplify_expr_floatbv.cpp [hunk 5]
```diff
@@ -142,7 +141,9 @@ bool simplify_exprt::simplify_floatbv_typecast(exprt &expr)
 {
   // These casts usually reduce precision, and thus, usually round.
 
-  assert(expr.operands().size()==2);
+  PRECONDITION(expr.id() == ID_floatbv_typecast);
+  DATA_INVARIANT(
+    expr.operands().size() == 2, "float typecast has two operands");
```
Comment (tautschnig) [inline src/util/simplify_expr_floatbv.cpp]: This function should take a `floatbv_typecast_exprt` as a parameter and should not be doing those checks in here.

**Review by tautschnig — CHANGES_REQUESTED** [review]
Let's try to do a bit more than the very mechanical cleanup.

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: a2c876d16fb60b0652ab1c5f41d56998f3ebdbb5).
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

File: src/util/simplify_expr_floatbv.cpp (hunk 5 above)
Comment (hannes-steffenhagen-diffblue) [reply src/util/simplify_expr_floatbv.cpp]: I agree that'd be much cleaner, but note that currently all of the `simplify_XXX` work like this (this is also because many of the simplify functions can take many different types of expressions). We could change that but if we were to do that I'd change all of them at once and that's out of scope for this PR IMHO (if we just changed the one the code would be quite inconsistent)

Comment (tautschnig) [reply src/util/simplify_expr_floatbv.cpp]: > [...] currently all of the [...]

Not quite true - there is sort of a gradual migration happening. Take a look at the declarations in `simplify_expr_class.h`.

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: a2c876d16fb60b0652ab1c5f41d56998f3ebdbb5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/83831905
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 9c79c23d13caf6732ae461fdb93065e10fbc95f5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84019963

File: src/util/simplify_expr_floatbv.cpp (hunk 5 above)
Comment (hannes-steffenhagen-diffblue) [reply src/util/simplify_expr_floatbv.cpp]: Hm... I don't really like that. It ends up violating structural invariants to still use the input expression as an out parameter, because most simplifications will "return" a different type than they were given... I find the idea of returning something that's _not_ a `plus_exprt` through a `plus_exprt` reference a bit disturbing.

Comment (tautschnig) [reply src/util/simplify_expr_floatbv.cpp]: See https://github.com/diffblue/cbmc/pull/2910#discussion_r215971488.

Comment (hannes-steffenhagen-diffblue) [reply src/util/simplify_expr_floatbv.cpp]: @tautschnig Can you take a look at https://github.com/diffblue/cbmc/pull/2939 and tell me if that's roughly in line with what you were thinking? Would you be OK with going ahead with this PR and doing further refactorings in one or more separate PRs?

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: b0b44fade96e6536db88ee41d7437087fb5b627d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85534606
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 42f660e92c2f59f17a8bb6a1ba484cc8b8eec7fe).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85532856

File: src/util/simplify_expr.cpp (hunk 1 above)
Comment (sonodtt) [reply src/util/simplify_expr.cpp]: removed.

File: src/util/simplify_expr.cpp (hunk 2 above)
Comment (sonodtt) [reply src/util/simplify_expr.cpp]: Refactored.

File: src/util/simplify_expr.cpp (hunk 3 above)
Comment (sonodtt) [reply src/util/simplify_expr.cpp]: Done.

File: src/util/simplify_expr_floatbv.cpp (hunk 5 above)
Comment (sonodtt) [reply src/util/simplify_expr_floatbv.cpp]: @tautschnig  I'm going to split this part out - and make a note for it to be considered in the context of #2939  and your remarks on https://github.com/diffblue/cbmc/pull/2910#discussion_r215971488

see 
https://github.com/diffblue/cbmc/pull/2939#issuecomment-424610272 that documents this

Comment (tautschnig) [reply src/util/simplify_expr_floatbv.cpp]: Sounds good!

File: src/util/simplify_expr.cpp [hunk 6]
```diff
@@ -1365,17 +1365,12 @@ bool simplify_exprt::simplify_update(exprt &expr)
...
-
-      std::size_t number=to_struct_type(value_ptr_type).
-        component_number(component_name);
-
-      CHECK_RETURN(number < value_ptr->operands().size());
-
-      value_ptr=&value_ptr->operands()[number];
+      auto &designator_as_struct_expr = to_struct_expr(*value_ptr);
+      value_ptr = &designator_as_struct_expr.component(component_name, ns);
+      CHECK_RETURN(value_ptr->id() != ID_nil);
```
Comment (tautschnig) [inline src/util/simplify_expr.cpp]: Use `value_ptr->is_not_nil()`

File: src/util/simplify_expr.cpp [hunk 7]
```diff
@@ -1365,17 +1365,12 @@ bool simplify_exprt::simplify_update(exprt &expr)
     {
       const irep_idt &component_name=
         e.get(ID_component_name);
-
-      if(!to_struct_type(value_ptr_type).
-         has_component(component_name))
+      const struct_typet &value_ptr_struct_type=to_struct_type(value_ptr_type);
```
Comment (tautschnig) [inline src/util/simplify_expr.cpp]: Nit pick: whitespace around `=`

File: src/util/simplify_expr.cpp [hunk 8]
```diff
@@ -1407,18 +1402,13 @@ bool simplify_exprt::simplify_object(exprt &expr)
   }
   else if(expr.id()==ID_typecast)
   {
-    const typet &op_type=ns.follow(expr.op0().type());
-
-    DATA_INVARIANT(
-      expr.operands().size() == 1,
-      "typecasts must have exactly one argument");
+    auto const& typecast_expr = to_typecast_expr(expr);
+    const typet &op_type=ns.follow(typecast_expr.op().type());
```
Comment (tautschnig) [inline src/util/simplify_expr.cpp]: Space around `=`

File: src/util/simplify_expr.cpp [hunk 9]
```diff
@@ -1429,10 +1419,10 @@ bool simplify_exprt::simplify_object(exprt &expr)
       // We do a bit of special treatment for (TYPE *)(a+(int)&o) and
       // (TYPE *)(a+(int)((T*)&o+x)), which are re-written to '&o'.
 
-      exprt tmp=expr.op0();
-      if(tmp.id()==ID_plus && tmp.operands().size()==2)
+      exprt casted_expr=typecast_expr.op();
```
Comment (tautschnig) [inline src/util/simplify_expr.cpp]: `const exprt &casted_expr = typecast_expr.op();`

File: src/util/simplify_expr.cpp [hunk 10]
```diff
@@ -1429,10 +1419,10 @@ bool simplify_exprt::simplify_object(exprt &expr)
...
+      if(casted_expr.id()==ID_plus && casted_expr.operands().size()==2)
```
Comment (tautschnig) [inline src/util/simplify_expr.cpp]: Add spacing around `==`

File: src/util/simplify_expr.cpp [hunk 11]
```diff
@@ -1429,10 +1419,10 @@ bool simplify_exprt::simplify_object(exprt &expr)
...
       {
-        exprt cand=tmp.op0().id()==ID_typecast?tmp.op0():tmp.op1();
+        exprt cand=casted_expr.op0().id()==ID_typecast?casted_expr.op0():casted_expr.op1();
```
Comment (tautschnig) [inline src/util/simplify_expr.cpp]: Spacing

File: src/util/simplify_expr_boolean.cpp [hunk 12]
```diff
@@ -213,15 +213,12 @@ bool simplify_exprt::simplify_not(exprt &expr)
...
-      "exists expression has two parameters");
-    exprt tmp;
-    tmp.swap(op);
-    expr.swap(tmp);
-    expr.id(ID_forall);
-    expr.op1().make_not();
-    simplify_node(expr.op1());
+    auto const& op_as_exists = to_exists_expr(op);
+    forall_exprt rewritten_op(op_as_exists.symbol()
+      , op_as_exists.where());
```
Comment (tautschnig) [inline src/util/simplify_expr_boolean.cpp]:
```
forall_exprt rewritten_op(
  op_as_exists.symbol(), not_exprt(op_as_exists.where()));
```
and remove the line below

File: src/util/simplify_expr.cpp [hunk 13]
```diff
@@ -1365,7 +1365,8 @@ bool simplify_exprt::simplify_update(exprt &expr)
     {
       const irep_idt &component_name=
         e.get(ID_component_name);
-      const struct_typet &value_ptr_struct_type=to_struct_type(value_ptr_type);
+      const struct_typet &value_ptr_struct_type =
+        to_struct_type(value_ptr_type);
```
Comment (tautschnig) [inline src/util/simplify_expr.cpp:1369]: Please squash with the previous commit, the clang-format changes are restricted to code that you have introduced in the preceding commit.

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 8734b32c850d0269f023929cd691f1333d84ab0d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86134560
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: cba81d933c1a9afbb5c834b2e27e6f9f2718b1bc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86174685

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 6c0ab09e74e6b6b577b6b4252e1cd51a56b00dfb).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86178555

**Comment by hannes-steffenhagen-diffblue** [comment]
@tautschnig I believe all of your comments (aside from changing the function signatures to be more strongly typed which will happen in another PR as previously discussed) have been addressed now.

**Review by tautschnig — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: a168b24b8da1a95938a4df6fc65510347fc51e42).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86316058
